### DISCUSSION TITLE
Brewtarget v3.0.0 is released

### DISCUSSION THREAD

**Discussion by matty0ung** [Brewtarget/brewtarget#660] [General]
Lots of updates, bug-fixes, new features and improvements.

<hr /><em>This discussion was created from the release <a href='https://github.com/Brewtarget/brewtarget/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by matty0ung** [comment]
Please report any bugs, in particular with the installation process.  We'll try to fix any install problems quickly.

**Reply by b-a-p** [reply]
Hi,
Quick feedback: Windows do not trust this file and suggests do not run it. I believe it makes sense to tweak the instalator.
Also I requested instalator to create desktop icon but can't find any after installation completed.
P.S.: Windows 11

**Reply by matty0ung** [reply]
Thanks for the feedback.  Can you give the exact text of the warning message?

**Reply by matty0ung** [reply]
Will look at the desktop icon issue too.

**Reply by mattiasmaahl** [reply]
The not trusting is due to the fact we don't have a trusted certificate to sign the installer. Don't remember the details but a signing cert has to be signed by a trusted provider and then used to sign the Msi. 

We don't so therefor the warning. You should be able to install anyways though?

**Comment by matty0ung** [comment]
I have a possible fix for the desktop icon on Windows (in as much as I've found some more CPack parameters that seem to control it).

I also want to bring in a fix for https://github.com/Brewtarget/brewtarget/issues/649 as the code there really isn't working.  (This is another follow-on fix from the new DB layer, a consequence of my misunderstanding about the responsibility split between WaterEditor and WaterDialog.  Fix is relatively straightforward and I can also, I hope, do the "missing" part of the radar chart for almost no work.)

**Comment by matty0ung** [comment]
Starting a list for 3.0.1:

- I have a possible fix for the desktop icon on Windows (in as much as I've found some more CPack parameters that seem to control it).

- I also want to bring in a fix for https://github.com/Brewtarget/brewtarget/issues/649 as the code there really isn't working.  (This is another follow-on fix from the new DB layer, a consequence of my misunderstanding about the responsibility split between WaterEditor and WaterDialog.  Fix is relatively straightforward and I can also, I hope, do the "missing" part of the radar chart for almost no work.)

- Database is going in the wrong directory, and both database and config+logs go in the wrong directory if you build against debug version of Qt -- see https://github.com/Brewtarget/brewtarget/issues/662.  Fix is small.

**Comment by brianbuccola** [comment]
Thanks for this! I was able to build and install on Linux without any problem.

However, it seems that the new version does not recognize any of my settings or recipes. It opens as if it were a totally new program being opened for the very first time.

When I downgrade to 2.3.1 and reopen brewtarget, I see all my settings and recipes.

Any tips? Do I need to "migrate" my old settings and recipes to the new program/database, and if so, how can I do that?

I'm running brewtarget on Arch Linux. For others who are too, my comment on [the AUR package](https://aur.archlinux.org/packages/brewtarget) describes how I updated the PKGBUILD to get brewtarget to build.

**Reply by matty0ung** [reply]
Ah, that's interesting.  I'm guessing the newer version of the program is looking in different folders for config and DB files.  That may well be my fault.  For Brewken, I did some simplification of the logic for where to look for these things and I might have merged too much of that logic back to Brewtarget.

I think the current default location for config and logs is `~/.config/brewtarget/` and `~/.local/share/brewtarget/` for the DB.  You could hopefully find the current 2.3.1 locations with the following commands:
```
echo $XDG_CONFIG_HOME
cd ~
find . -type f -name brewtarget.conf
find . -type f -name brewtarget.log
find . -type f -name database.sqlite
```

**Reply by matty0ung** [reply]
It should then be a case of copying the old files to the new directories and hopefully everything just works (TM).  Obviously make sure you've got backups first, and let us know how you get on.  If need be I can add an extra search path in the 3.0.1 patch release.

**Reply by brianbuccola** [reply]
Ever since v2.3, my config and database files have all been in `~/.config/brewtarget` (where they still are):

```
% tree ~/.config/brewtarget
/home/brian/.config/brewtarget
├── brewtarget.conf
├── brewtarget_log.txt
├── database.sqlite
└── obsolete

1 directory, 3 files
```

(Nothing in `~/.local/share`.)

But now I also see a `brewtarget-debug` directory in both `~/.config` and `~/.local/share`:

```
% tree ~/.config/brewtarget-debug
/home/brian/.config/brewtarget-debug
├── brewtarget.conf
└── brewtarget.log

0 directories, 2 files
%
% tree ~/.local/share/brewtarget-debug
/home/brian/.local/share/brewtarget-debug
└── database.sqlite

0 directories, 1 file
```

Is this where brewtarget v3.0.0 is looking for the config and database? (`debug` suggests not...) Or where should I copy my old config and database to?

(Thanks for the help.)

**Reply by matty0ung** [reply]
What are the timestamps on the files?
If the brewtarget-debug stuff is the newest then I've possibly got a build parameter wrong.

**Reply by brianbuccola** [reply]
Yes, I just now moved all existing brewtarget files to `/tmp`, opened brewtarget v3.0.0, closed it, and now all I see is:

```
% tree -D ~/.config/brewtarget-debug
[Sep 25 11:22]  /home/brian/.config/brewtarget-debug
└── [Sep 25 11:22]  brewtarget.conf

0 directories, 1 file
%
% tree -D ~/.local/share/brewtarget-debug
[Sep 25 11:22]  /home/brian/.local/share/brewtarget-debug
└── [Sep 25 11:22]  database.sqlite

0 directories, 1 file
```

There is no `~/.config/brewtarget`:

```
% find . -type f -name "brewtarget.conf"
./.config/brewtarget-debug/brewtarget.conf
```

If I copy my old `brewtarget.conf` and `database.sqlite` to their respective `brewtarget-debug` directory, and open brewtarget, then I see all my recipes. However, my config options do not seem to be respected. I haven't checked them all, but for example my config specifies US units but brewtarget shows metric/SI by default.

So to recap:

1. brewtarget v3.0.0 is using `brewtarget-debug` as the default config/database directory.
2. There seems to be some incompatibility between v2.3.1 config file and v3.0.0. Maybe the names of parameters have changed?

**Reply by matty0ung** [reply]
Thanks for the info and your patience on this.  I've raised https://github.com/Brewtarget/brewtarget/issues/662 and will have a look at making a fix to go in a 3.0.1 patch release.

**Reply by matty0ung** [reply]
I've pushed a potential fix for this into https://github.com/Brewtarget/brewtarget/pull/661.

There are two parts to the fix.  Firstly, I realised that the reason you're getting the `-debug` on directory names was most probably because you're compiling from source against the debug version of Qt.  I've changed the Brewtarget source code so it doesn't add this `-debug` for debug Qt, as I don't think it's helpful.

Secondly, I've changed the code so it puts the database in the "config" directory (`QStandardPaths::AppConfigLocation`) rather than the "data" directory (`QStandardPaths::AppDataLocation`).  Brewken uses the latter on the grounds that it's arguably more standard, but we should retain everything in the same directory for Brewtarget as that keeps things working for folks with existing installs.  (Both directories are guaranteed writeable, so it's not a problem to use the "config" one for the database.)

**Comment by matty0ung** [comment]
OK, so I've pushed out some fixes in https://github.com/Brewtarget/brewtarget/releases/tag/v3.0.1

As ever, please don't hesitate to let us know any problems you encounter.

**Comment by matty0ung** [comment]
Archiving this now, since we're on the 4.0 release.
